### ISSUE TITLE
Integrate CORE-MATH implementations into openlibm

### ISSUE THREAD

**Issue by ViralBShah** [JuliaMath/openlibm#260]
Discussed in: https://github.com/JuliaMath/openlibm/issues/212#issuecomment-1077587522

**Comment by zimmermann6** [comment]
The following reference might be relevant : https://hal.inria.fr/hal-03721525. The CORE-MATH developers are ready to help with any issue found during this integration.

**Comment by JeffreySarnoff** [comment]
lean on me

**Comment by zimmermann6** [comment]
yes CORE-MATH provides a testsuite. For each function, say cbrt, you can do ```./check.sh cbrt``` which will check all worst cases for all four rounding modes. For a few functions, you can also do ```./check.sh --special cbrt```, which will check special cases (for cbrt exact and midpoint cases).

**Comment by zimmermann6** [comment]
Hi Viral,

yes CORE-MATH does provide a testsuite.

For univariate single precision functions, all possible inputs (about 2^32)
are checked for correct rounding against MPFR, for all 4 rounding modes:

$ ./check.sh --exhaustive exp10f
Running exhaustive check in --rndn mode...
...

For larger precision or bivariate functions, some tests are done for
hard-to-round cases, still against MPFR and for all 4 rounding modes:

$ ./check.sh exp
Running worst cases check in --rndn mode...
68059 tests passed, 0 failure(s)
Running worst cases check in --rndz mode...
68059 tests passed, 0 failure(s)
Running worst cases check in --rndu mode...
68059 tests passed, 0 failure(s)
Running worst cases check in --rndd mode...
68059 tests passed, 0 failure(s)

Best regards,
Paul

**Comment by ViralBShah** [comment]
I would love to see this happen, but the core-math codebase is not something I am able to easily build or understand for integrating (in the little time I spent looking at it). Would be great if someone more familiar wants to start off with a couple of functions and see how this goes.

**Comment by ValeZAA** [comment]
> I would love to see this happen, but the core-math codebase is not something I am able to easily build or understand for integrating (in the little time I spent looking at it). Would be great if someone more familiar wants to start off with a couple of functions and see how this goes.

CORE-MATH is now integrated partially into glibc.

**Comment by JeffreySarnoff** [comment]
"CORE-MATH is now integrated partially into glibc"
What part, and is it the default for that part?

**Comment by zimmermann6** [comment]
Hi Viral, I can help. I suggest starting with the single precision exponential function https://gitlab.inria.fr/core-math/core-math/-/blob/master/src/binary32/exp/expf.c. This code is self-content, thus should be easy to integrate.

Jeffrey, the functions integrated into GNU libc are (see CORE-MATH web page): "seven binary64 functions are integrated into the GNU libc up from release 2.43 (acosh, asinh, atanh, erf, erfc, lgamma, tgamma), and the following binary32 functions are integrated into the GNU libc up from release 2.42: acosf, acoshf, asinf, asinhf, atanf, atanhf, cbrtf, coshf, erff, erfcf, expm1f, lgammaf, log10f, log1pf, sinhf, tanf, tanhf, tgammaf, acospif, asinpif, atanpif, cospif, exp10m1f, exp2m1f, log10p1f, log2p1f, sinpif, tanpif, atan2f, atan2pif"

**Comment by zimmermann6** [comment]
sorry I forgot to answer this "What part, and is it the default for that part?".

For the integrated functions, the previous GNU libc code was replaced by the correctly-rounded code.
Thus it is indeed the default.

**Comment by ViralBShah** [comment]
I thought it would be interesting to let claude code lose on this and let it integrate core-math routines into openlibm. It would be interesting to pick which ones to go with, since the performance difference is substantial in some cases.

@zimmermann6 I'll start with your list.

**Comment by ValeZAA** [comment]
> sorry I forgot to answer this "What part, and is it the default for that part?".
> 
> For the integrated functions, the previous GNU libc code was replaced by the correctly-rounded code. Thus it is indeed the default.

Sinf is still not integrated, is it because it is slightly slower than what glibc already has?

BTW, Android uses fdlibm, that would be a very good vector to be fixed next. https://android.googlesource.com/platform/external/fdlibm/

In particular Chrome uses fdlibm on Android, you see. What is worse on Windows also Chrome uses fdlibm becuse they want the results to be the same across platforms.

**Comment by zimmermann6** [comment]
sinf is being integrated in GNU libc : https://sourceware.org/pipermail/libc-alpha/2026-March/176259.html

They indeed integrate first routines that are clearly faster
